NLSY/nlsy_lib.R

```diff
@@ -241,6 +241,121 @@ nlsy_get_highest_grade_completed_df = function()
     return(data)
 }
 
+#' Returns marriage status the last month of a given year
+nlsy_get_marstat_df = function()
+{
+    data = nlsydf |>
+        select(
+            id =        "PUBID_1997",
+            starts_with('MAR_STATUS_')
+        ) |>
+        pivot_longer(
+            starts_with('MAR_STATUS_'),
+            names_to=c('year', 'month'),
+            names_pattern='MAR_STATUS_(\\d\\d\\d\\d)\\.(\\d\\d)_XRND',
+            values_to='mar_status'
+        ) |>
+        mutate(
+            year=as.integer(year),
+            month=as.integer(month),
+            single=as.integer(mar_status == 'Never Married, Not Cohabitating'),
+            married=as.integer(mar_status == 'Married'),
+            divorced=as.integer(mar_status == 'Divorced'),
+            widowed=as.integer(mar_status == 'Widowed')
+        )
+
+    data = data |>
+        filter(month == 12) |>
+        group_by(id, year) |>
+        summarise(
+            allna = all(is.na(mar_status)),
+            single=any(single, na.rm=TRUE),
+            married=any(married, na.rm=TRUE),
+            divorced=any(divorced, na.rm=TRUE),
+            widowed=any(widowed, na.rm=TRUE),
+        ) |>
+        filter(!allna) |>
+        select(-allna)
+
```

Comment:
Before we return, we should check if we have values for all people and all years. If some people are completely missing, we should count how many. If some years are missing for a person, we should add observations for them with `complete` and fill their values with the previous year value with `fill` like you did in `nlsy_get_biochild_df`.

Reply:
I think I understand how to use `fill()` a bit better, which addressed some of the missings I was seeing for the month of December. 

After using fill, there's a few IDs with missings across all of the years. By my count, there's about 203 out of the 8984 respondents in nlsydf. For now, I have replaced those missings with "never married, not cohabitating" (but I'm happy to revert this back and just delete those entries). These new edits get rid of all missing values.

Reply:
Depending on the patterns of missing data, we could apply `fill` twice. First, before calling `filter(month==12)`, we can fill months within a year. This would be useful if some December values are missing but other values in the same year for the same person are present. To do this, we should `group_by(id, year)` and call `fill`. Second, after we drop all months but December, and we call `complete`, we `group_by(id)` and call `fill` again to fill in missing years.

Reply:
Thanks! After doing the double fill, the number of missings (before the `case_when`) remains the same. The remaining missings appear to be people who have no reported relationship status for any year/month in the data. I have kept the previous line of code that replaced all outstanding zeroes after the fills with: 
`mutate(mar_status = case_when(
            is.na(mar_status) ~ 'Never Married, Not Cohabitating',
            TRUE ~ mar_status)`

Reply:
Flagging that the `complete` function is really slow (likely because it has to operate row-wise). Works totally fine, but we may want to look at a `data.table` version of `complete` if we're finding that it gets annoying to call repeatedly.

Reply:
Thanks @storresrod! @judah-axelrod, you're right, complete is pretty slow. How easy/difficult would it be to replace it with data.table? I feel that this is still not a critical issue. I try to mitigate it in Quarto documents by using caching.